src/main/java/com/oath/halodb/HaloDBInternal.java

```diff
@@ -281,6 +286,66 @@ int get(byte[] key, ByteBuffer buffer) throws IOException {
         }
     }
 
+    public boolean takeSnapShot() {
+
+        logger.info("Start generating the snapshot");
+
+        File[] filesToCopy = dbDirectory.getPath().toFile()
+            .listFiles(file -> {
+                return (Constants.DATA_FILE_PATTERN.matcher(file.getName()).matches()
+                    || Constants.INDEX_FILE_PATTERN.matcher(file.getName()).matches()
+                    || Constants.TOMBSTONE_FILE_PATTERN.matcher(file.getName()).matches());
+            } );
+
+        logger.info("file number need to be copied: {}", filesToCopy.length);
+
+        try {
+            Path dbDirectoryPath = dbDirectory.getPath();
+            File snapShotDir = Paths.get(dbDirectoryPath.toFile().getAbsolutePath(), "snapshot").toFile();
+            // Only support one snapshot now
+            // TODO: support multiple snapshots if needed
+            if (snapShotDir.exists()) {
+                logger.warn("The snapshot dir is already existed. Delete the old one.");
+                FileUtils.deleteDirectory(snapShotDir);
+            }
+
+            FileUtils.createDirectoryIfNotExists(snapShotDir);
+            logger.info("Created directory for snapshot {}", snapShotDir.toString());
+
+            for (File file : filesToCopy) {
+                if (file.exists()) {
+                    Path dest = Paths.get(snapShotDir.getAbsolutePath(), file.getName());
+                    logger.debug("Create links from file {} to {}", file.getName(), dest.toFile().getAbsoluteFile());
+                    Files.createLink(dest, file.toPath());
+                } else {
+                    logger.warn("File {} maybe deleted by compaction job", file.getName());
+                }
+            }
+        } catch (IOException e) {
+            logger.warn("IO exception in copy, snapshot failure");
+
+            return false;
+        }
+
+        // TODO: Need solutions to handle the file pieces after the rollover
+        // especially there are frequently rollovers
+        writeLock.lock();
```

Comment:
It seems that ```writeLock``` is enough for the whole procedure. The workflow should be:
1. Pause compaction
2. writeLock.lock();
3. put current write file to compaction list and then rollover current write file
4. Rollover current tombstone file
5. Construct snapshot file list by **excluding** current write file and tombstone file
6. Put current compaction file to compaction list and rollover compaction file
7.  writeLock.unlock();
8. Create file links for snapshot
9. Resume compaction job

Reply:
Yes, looks writeLock should be enough for tombstone and currentWriteFile.